File: Visual Basic .NET/Selection/CustomerConverter.vb
'
' Xceed DataGrid for WPF - SAMPLE CODE - Selection Sample Application
' Copyright (c) 2007-2024 Xceed Software Inc.
' 
' [CustomerConverter.vb]
'
' This file demonstrates how to create a converter to display a company name
' followed by an optional contact name between parentheses.
' 
' This file is part of the Xceed DataGrid for WPF product. The use 
' and distribution of this Sample Code is subject to the terms 
' and conditions referring to "Sample Code" that are specified in 
' the XCEED SOFTWARE LICENSE AGREEMENT accompanying this product.
'

Imports System
Imports System.Data
Imports System.Globalization
Imports System.Windows
Imports System.Windows.Data

Namespace Xceed.Wpf.DataGrid.Samples.Selection
  Public Class CustomerConverter
    Implements IValueConverter

    Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.Convert
      If value Is Nothing Then
        Return Nothing
      End If

      'By default, the parameter will contain the ItemsSource define on the ForeignKeyConfiguration when this converter is set as a ForeignKeyValueConverter.
      Dim customers As DataView = TryCast(parameter, DataView)
      If customers Is Nothing Then
        Return Nothing
      End If

      customers.Sort = "CustomerID"

      Dim index As Integer = customers.Find(value)
      Dim dataRow As DataRowView = customers(index)

      Return dataRow("CompanyName") & " - " & dataRow("ContactName")
    End Function

    Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
      Return Nothing
    End Function
  End Class

  Public Class CustomerMultiConverter
    Implements IMultiValueConverter

    Public Function Convert(ByVal values As Object(), ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IMultiValueConverter.Convert
      If (values(0) Is DependencyProperty.UnsetValue) Then
        values(0) = String.Empty
      End If

      If (values(1) Is DependencyProperty.UnsetValue) Then
        values(1) = String.Empty
      End If

      Dim name As String = CType(values(0), String)

      If Not String.IsNullOrEmpty(values(1)) Then
        name = name + " - " + CType(values(1), String)
      End If

      Return name
    End Function

    Public Function ConvertBack(ByVal value As Object, ByVal targetTypes As Type(), ByVal parameter As Object, ByVal culture As CultureInfo) As Object() Implements IMultiValueConverter.ConvertBack
      Throw New NotImplementedException()
    End Function

  End Class

  Public Class CustomerForeignKeyConverter
    Inherits DataTableForeignKeyConverter
    'Only the GetValueFormKey method overload with the DataGridForeignKeyDescription parameter needs to be overridden,
    'for the base class DataTableForeignKeyConverter already provides an implementation of the other two overridable methods.
    Public Overrides Function GetValueFromKey(ByVal key As Object, ByVal description As DataGridForeignKeyDescription) As Object
      If key Is Nothing Then
        Return Nothing
      End If

      Dim dataView As DataView = TryCast(description.ItemsSource, DataView)
      If dataView IsNot Nothing Then
        dataView.Sort = description.ValuePath

        Dim index As Integer = dataView.Find(key)
        Dim dataRow As DataRowView = dataView(index)

        'Return a value built in this order, so sorting is done on company name, then contact name.
        Return dataRow("CompanyName") & " - " & dataRow("ContactName")
      End If

      Return key
    End Function
  End Class
End Namespace

File: CSharp/SpannedCells/CustomerConverter.cs
/*
 * Xceed DataGrid for WPF - SAMPLE CODE - Spanned Cells Sample Application
 * Copyright (c) 2007-2024 Xceed Software Inc.
 * 
 * [CustomerConverter.cs]
 *  
 * This file demonstrates how to create a converter to display a company name
 * followed by an optional contact name between parentheses.
 * 
 * This file is part of the Xceed DataGrid for WPF product. The use 
 * and distribution of this Sample Code is subject to the terms 
 * and conditions referring to "Sample Code" that are specified in 
 * the XCEED SOFTWARE LICENSE AGREEMENT accompanying this product.
 *
 */
using System;
using System.Data;
using System.Globalization;
using System.Windows;
using System.Windows.Data;

namespace Xceed.Wpf.DataGrid.Samples.SpannedCells
{
  public class CustomerConverter : IValueConverter
  {
    object IValueConverter.Convert( object value, Type targetType, object parameter, System.Globalization.CultureInfo culture )
    {
      if( value == null )
        return null;

      //By default, the parameter will contain the ItemsSource define on the ForeignKeyConfiguration when this converter is set as a ForeignKeyValueConverter.
      var customers = parameter as DataView;
      if( customers == null )
        return null;

      customers.Sort = "CustomerID";

      var index = customers.Find( value );
      var dataRow = customers[ index ];

      return dataRow[ "CompanyName" ] + " - " + dataRow[ "ContactName" ];
    }

    object IValueConverter.ConvertBack( object value, Type targetType, object parameter, System.Globalization.CultureInfo culture )
    {
      return null;
    }
  }

  public class CustomerMultiConverter : IMultiValueConverter
  {
    public object Convert( object[] values, Type targetType, object parameter, CultureInfo culture )
    {
      if( values[ 0 ] == DependencyProperty.UnsetValue )
        values[ 0 ] = string.Empty;

      if( values[ 1 ] == DependencyProperty.UnsetValue )
        values[ 1 ] = string.Empty;

      string customerName = ( string )values[ 0 ];

      if( !string.IsNullOrEmpty( values[ 1 ] as string ) )
        customerName += " - " + ( string )values[ 1 ];

      return customerName;
    }

    public object[] ConvertBack( object value, Type[] targetTypes, object parameter, CultureInfo culture )
    {
      return null;
    }
  }

  public class CustomerForeignKeyConverter : DataTableForeignKeyConverter
  {
    //Only the GetValueFormKey method overload with the DataGridForeignKeyDescription parameter needs to be overridden,
    //for the base class DataTableForeignKeyConverter already provides an implementation of the other two overridable methods.
    public override object GetValueFromKey( object key, DataGridForeignKeyDescription description )
    {
      if( key == null )
        return null;

      var dataView = description.ItemsSource as DataView;
      if( dataView != null )
      {
        dataView.Sort = description.ValuePath;

        var index = dataView.Find( key );
        var dataRow = dataView[ index ];

        //Return a value built in this order, so sorting is done on company name, then contact name.
        return dataRow[ "CompanyName" ] + " - " + dataRow[ "ContactName" ];
      }

      return key;
    }
  }
}
